extendr-api/tests/graphics_tests.rs

```diff
@@ -92,3 +96,236 @@ fn graphics_test() {
+            })
+            .collect::<Vec<String>>()
+            .join(", ");
+
+        writeln!(*self.canvas, "path coords=[{coords}] winding={winding}").unwrap();
+    }
+
+    fn text(
+        &mut self,
+        pos: (f64, f64),
+        str: &str,
+        rot: f64,
+        hadj: f64,
+        _: R_GE_gcontext,
+        _: DevDesc,
+    ) {
+        let (x, y) = pos;
+        writeln!(
+            *self.canvas,
+            "text pos=({x:.1}, {y:.1}) str='{str}' rot={rot:.1} hadj={hadj:.1}"
+        )
+        .unwrap();
+    }
+
+    fn raster<T: AsRef<[u32]>>(
+        &mut self,
+        raster: Raster<T>,
+        pos: (f64, f64),
+        size: (f64, f64),
+        rot: f64,
+        interpolate: bool,
+        _: R_GE_gcontext,
+        _: DevDesc,
+    ) {
+        let (x, y) = pos;
+        let (width, height) = size;
+        let Raster { pixels, width: w } = raster;
+        let pixels_str = pixels
+            .as_ref()
+            .iter()
+            .map(|&p| format!("{p}"))
+            .collect::<Vec<String>>()
+            .join("|");
+
+        writeln!(
+                *self.canvas,
+                "raster {pixels_str} w={w} pos=({x:.1}, {y:.1}) size=({width:.1}, {height:.1}) rot={rot:.1} interpolate={interpolate}"
+            )
+            .unwrap();
+    }
+}
+
+#[test]
+fn graphics_device_test() {
+    test! {
+        let mut value = 0.0;
+        let mut last_mode = 0;
+        let mut closed = false;
+        let mut canvas = String::new();
+
+        let device_driver = TestDevice {
+            last_mode: &mut last_mode,
+            value: &mut value,
+            canvas: &mut canvas,
+            closed: &mut closed,
+        };
+
+        let device_descriptor = DeviceDescriptor::new();
+        let device = device_driver.create_device::<TestDevice>(device_descriptor, "test device");
+
+        let gc = Context::from_device(&device, Unit::Device);
+
+        // if activate() is invoked, value should be 100.0
+        assert_eq!(value, 100.0);
+        assert!(!closed);
+
+        device.mode_on().unwrap();
+
+        // if mode() is invoked, value and last_mode should be updated
+        assert_eq!(last_mode, 1);
+        assert_eq!(value, 101.0);
+
+        device.mode_off().unwrap();
+
+        // if mode() is invoked, value and last_mode should be updated
+        assert_eq!(last_mode, 0);
+        assert_eq!(value, 102.0);
+
+        device.clip((1.1, 2.2), (3.3, 4.4), &gc);
+        device.circle((1.1, 2.2), 3.3, &gc);
+        device.line((1.1, 2.2), (3.3, 4.4), &gc);
+        device.rect((1.1, 2.2), (3.3, 4.4), &gc);
+
+        // // TODO: uncomment the following lines when https://github.com/extendr/extendr/issues/370 is solved.
+        // device.polyline([(0.0, 0.0), (0.0, 2.0)], &gc);
+        // device.polygon([(0.0, 0.0), (1.0, 2.0), (2.0, 0.0)], &gc);
+        // device.path([[(0.0, 0.0), (1.0, 0.0), (1.0, 1.0), (0.0, 1.0)], [(0.3, 0.0), (0.3, 0.3), (0.7, 0.3), (0.3, 0.7)]], true, &gc);
```

Comment:
For some reason, these ones don't work. I'll uncomment here when https://github.com/extendr/extendr/issues/370 is resolved.